File: N/python/code.py
from node import Node

# Comment it before submitting
# class Node:
#     def __init__(self, val):
#         self.val = val
#         self.neighbours = []

visited_nodes = dict()

def cloneGraph(node) -> Node:
    # Your code
    # “ヽ(´▽｀)ノ”
    if node.val in visited_nodes:
        return visited_nodes[node.val]

    new_head = Node(node.val)
    visited_nodes[new_head.val] = new_head
    for n in node.neighbours:
        new_head.neighbours.append(cloneGraph(n))

    return new_head

File: N/cpp/code.cpp
#include "solution.h"
#include <vector>

// Comment it before submitting
// struct Node {
//     int val;
//     std::vector<Node*> neighbours;
//     Node(int val_) {
//         val = val_;
//         neighbours = {};
//     }
// };


Node* cloneGraph(Node* node) {
    static std::vector<Node*> visited;

    for (int i = 0; i < visited.size(); ++i) {
        if (visited[i]->val == node->val)
            return visited[i];
    }

    Node* new_node = new Node(node->val);
    visited.push_back(new_node);
    for (int i = 0; i < node->neighbours.size(); ++i) {
        cloneGraph(node->neighbours[i]);
    }

    return new_node;
}

// int main () {
//     return 0;
// }
